src/cpp/server/external_connection_acceptor_impl.cc

```diff
@@ -0,0 +1,90 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "src/cpp/server/external_connection_acceptor_impl.h"
+
+#include <memory>
+
+#include <grpcpp/server_builder_impl.h>
+#include <grpcpp/support/channel_arguments.h>
+
+namespace grpc_impl {
+namespace {
+class InternalAcceptor : public grpc::ExternalConnectionAcceptor {
+ public:
+  explicit InternalAcceptor(
+      std::shared_ptr<ExternalConnectionAcceptorImpl> impl)
+      : impl_(std::move(impl)) {}
+  void HandleNewConnection(NewConnectionParameters* p) override {
+    impl_->HandleNewConnection(p);
+  }
+
+ private:
+  std::shared_ptr<ExternalConnectionAcceptorImpl> impl_;
+};
+}  // namespace
+
+ExternalConnectionAcceptorImpl::ExternalConnectionAcceptorImpl(
+    const grpc::string& name,
+    ServerBuilder::experimental_type::ExternalConnectionType type,
+    std::shared_ptr<ServerCredentials> creds)
+    : name_(name), creds_(std::move(creds)) {
+  GPR_ASSERT(type == ServerBuilder::experimental_type::ExternalConnectionType::
+                         CONNECTION_FROM_FD);
+}
+
+std::unique_ptr<grpc::ExternalConnectionAcceptor>
```

Comment:
can return a shared ptr here to avoid wrapping a shared pointer in a unique ptr. avoids an allocation